genienlp/run_generation.py

```diff
@@ -201,76 +231,208 @@ def sample_sequence(model, length, context, position_ids, num_samples=1, tempera
-    'TIME_1': {'forward': '2p.m.', 'back': ['2 pm', '2pm', '2:00 pm', '2:00pm', '2p.m.', '2 p.m.', '2:00 p.m.', '2:00']},
-    'EMAIL_ADDRESS_0': {'forward': 'e1@example.com'},
-    'EMAIL_ADDRESS_1': {'forward': 'e2@example.com'},
-    'URL_0': {'forward': 'my1site.com'},
-    'URL_1': {'forward': 'my2site.com'},
-    'DATE_0': {'forward': '5-6-2015', 'back': ['5-6-2015']},
-    'DATE_1': {'forward': '8-3-2016', 'back': ['8-3-2016']},
-    'CURRENCY_0': {'forward': '$12', 'back': ['$12', 'twelve dollars', '12 dollars', '$ 12', '$ 12.00', '12.00', '12']},
-    'CURRENCY_1': {'forward': '$13', 'back': ['$13', 'thirteen dollars', '13 dollars', '$ 13', '$ 13.00', '13.00', '13']},
-    'NUMBER_0': {'forward': '2', 'back': ['2', 'two']},
-    'NUMBER_1': {'forward': '3', 'back': ['3', 'three']},
-    'DURATION_0': {'forward': '5 weeks', 'back': ['5 weeks', 'five weeks']},
-    'DURATION_1': {'forward': '6 weeks', 'back': ['6 weeks', 'six weeks']},
-    'LOCATION_0': {'forward': 'locatio1n', 'back': ['locatio1n', 'locat1n']},
-    'LOCATION_1': {'forward': 'locatio2n', 'back': ['locatio2n', 'locat2n']},
-    'PHONE_NUMBER_0': {'forward': '888-8888'},
-    'PHONE_NUMBER_1': {'forward': '777-8888'}
-}
+    return generated, generated_logits
+
+
+special_pattern_mapping = [
+    SpecialTokenMap('PHONE_NUMBER_([0-9]+)', ['888-8888', '777-8888']),
+    SpecialTokenMap('NUMBER_([0-9]+)', ['2', '3'], [['2', 'two'], ['3', 'three']]),
+    SpecialTokenMap('PATH_NAME_([0-9]+)', ['my1folder', 'my2folder']),
+    SpecialTokenMap('TIME_([0-9]+)', ['1p.m.', '2p.m.'], [['1 pm', '1pm', '1:00 pm', '1:00pm', '1p.m.', '1 p.m.', '1:00 p.m.', '1:00', 'one o\'clock', 'one'],
+                                                            ['2 pm', '2pm', '2:00 pm', '2:00pm', '2p.m.', '2 p.m.', '2:00 p.m.', '2:00', 'two o\'clock', 'two']]),
+    SpecialTokenMap('EMAIL_ADDRESS_([0-9]+)', ['e1@example.com', 'e2@example.com']),
+    SpecialTokenMap('URL_([0-9]+)', ['my1site.com', 'my2site.com']),
+    SpecialTokenMap('DATE_([0-9]+)', ['5-6-2015', '8-3-2016']),
+    SpecialTokenMap('CURRENCY_([0-9]+)', ['$12', '$13'], [['$12', 'twelve dollars', '12 dollars', '$ 12', '$ 12.00', '12.00', '12'], 
+                                                          ['$13', 'thirteen dollars', '13 dollars', '$ 13', '$ 13.00', '13.00', '13']]),
+    SpecialTokenMap('DURATION_([0-9]+)', ['5 weeks', '6 weeks'], [['5 weeks', 'five weeks'], ['6 weeks', 'six weeks']]),
+    SpecialTokenMap('LOCATION_([0-9]+)', ['locatio1n', 'locatio2n'], [['locatio1n', 'locat1n'], ['locatio2n', 'locat2n']]),
+    SpecialTokenMap('QUOTED_STRING_([0-9]+)', lambda x: 'Chinese', lambda x: ['Chinese', 'chinese']), # TODO change to be more general than cuisine
+    SpecialTokenMap('GENERIC_ENTITY_uk.ac.cam.multiwoz.Restaurant:Restaurant_([0-9]+)', ["restaurant1", "restaurant2", "restaurant3"]) # TODO the only reason we can get away with this unnatural replacement is that actual backward is not going to be called for this
```

Comment:
Do you have a plan for the TODOs here? I feel like we'll soon forget about this.